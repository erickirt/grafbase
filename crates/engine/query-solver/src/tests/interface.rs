use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/simple-interface-object/a")
    B @join__graph(name: "b", url: "http://localhost:4200/simple-interface-object/b")
    C @join__graph(name: "c", url: "http://localhost:4200/simple-interface-object/c")
}

type Admin implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    isActive: Boolean!
    name: String! @join__field
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
    @join__type(graph: C)
{
    users: [NodeWithName!]! @join__field(graph: A)
    anotherUsers: [NodeWithName] @join__field(graph: B)
    accounts: [Account] @join__field(graph: B)
}

type Regular implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    name: String! @join__field
    isActive: Boolean! @join__field
}

type User implements NodeWithName
    @join__implements(graph: A, interface: "NodeWithName")
    @join__type(graph: A, key: "id")
{
    id: ID!
    name: String
    age: Int
    username: String @join__field
}

interface Account
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
    @join__type(graph: C, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String! @join__field(graph: B)
    isActive: Boolean! @join__field(graph: C)
}

interface NodeWithName
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String @join__field(graph: A)
    username: String @join__field(graph: B)
}
"#;

#[test]
fn interface_field_providing_object_field() {
    // age is coming from subgraph A, but needs User.id for this. `anotherUsers` returns an
    // interface though, so need to retrieve the `NodeWithName.id` as an alternative for `User.id`
    assert_solving_snapshots!(
        "interface_field_providing_object_field",
        SCHEMA,
        r#"
        query {
          anotherUsers {
            ... on User {
              age
            }
          }
        }
        "#
    );
}

#[test]
fn entity_interface_field() {
    // username  comes from the subgraph B and requires the NodeWithName.id
    assert_solving_snapshots!(
        "entity_interface_field",
        SCHEMA,
        r#"
        query {
          users {
            ... on User {
              age
              username
            }
          }
        }
        "#
    );
}
