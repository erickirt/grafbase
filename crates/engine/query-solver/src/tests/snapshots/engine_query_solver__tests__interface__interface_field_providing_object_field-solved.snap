---
source: crates/engine/query-solver/src/tests/interface.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"User.age\", color=forestgreen ]\n    2 [ label = \"Root#b\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    3 [ label = \"anotherUsers#b\", shape=box, color=dodgerblue, color=forestgreen ]\n    4 [ label = \"FedEntity#a\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    5 [ label = \"age#a\", shape=box, color=dodgerblue, color=forestgreen ]\n    6 [ label = \"*NodeWithName.id\", color=forestgreen ]\n    7 [ label = \"id#b\", shape=box, color=dodgerblue, color=forestgreen ]\n    0 -> 2 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    2 -> 3 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    3 -> 4 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    4 -> 5 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    5 -> 1 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    3 -> 7 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    7 -> 6 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "User.age", steiner=1 ]
    2 [ label = "Root#b", steiner=1 ]
    3 [ label = "anotherUsers#b", steiner=1 ]
    4 [ label = "FedEntity#a", steiner=1 ]
    5 [ label = "age#a", steiner=1 ]
    6 [ label = "*NodeWithName.id", steiner=1 ]
    7 [ label = "id#b", steiner=1 ]
    0 -> 2 [ cost=1, steiner=1]
    2 -> 3 [ cost=0, steiner=1]
    3 -> 4 [ cost=1, steiner=1]
    4 -> 5 [ cost=0, steiner=1]
    5 -> 1 [ cost=0, steiner=1]
    3 -> 7 [ cost=0, steiner=1]
    7 -> 6 [ cost=0, steiner=1]
}
